cpp/src/arrow/array/data.h

```diff
@@ -434,6 +435,23 @@ struct ARROW_EXPORT ArraySpan {
     return GetValues<T>(i, this->offset);
   }
 
+  // Access a buffer's data as a span
+  template <typename T>
+  util::span<const T> GetSpan(int i, int64_t length) const {
+    int buffer_length = static_cast<size_t>(buffers[i].size) / sizeof(T);
```

Comment:
`int` might not be enough to fit the number of elements in the buffer, so this is better done with `int64_t` on all operands and the result:

```cpp
const int64_t buffer_length = buffers[i].size / static_cast<int64_t>(sizeof(T));
```

You can cast it to `size_t` later if needed.